### PR TITLE
fixed memory problem in tst_h_dimscales

### LINKED ISSUE #244: Test failure building against hdf5 1.10.0
> Please provide as much of the following information as you can, as applicable to the issue being reported.  Naturally, not all information is relevant to every issue, but the more information we have to start, the better!
### Environment Information

> Feel free to skip this if the issue is related to documentation, a feature request, or general discussion.
- What platform are you using? (please provide specific distribution/version in summary)
  - [X ] Linux
  - [ ] Windows
  - [ ] OSX
  - [ ] Other
  - [ ] NA
- 32 and/or 64 bit?
  - [ X] 32-bit
  - [ X] 64-bit
- What build system are you using?
  - [ X] autotools (`configure`)
  - [ ] cmake
- Can you provide a sample netCDF file or `C` code to recreate the issue?
  - [ X] Yes (please attach to this issue, thank you!)
  - [ ] No
  - [ ] Not at this time
### Summary of Issue

Building netcdf 4.4.0 against hdf5 1.10.0 I get the following test failure:

```
 cat build/h5_test/test-suite.log
==========================================
   netCDF 4.4.0: h5_test/test-suite.log
==========================================
# TOTAL: 26
# PASS:  25
# SKIP:  0
# XFAIL: 0
# FAIL:  1
# XPASS: 0
# ERROR: 0
.. contents:: :depth: 2
FAIL: tst_h_dimscales
=====================
HDF5-DIAG: Error detected in HDF5 (1.10.0) thread 0:
  #000: ../../src/H5Gdeprec.c line 1092 in H5Gget_objtype_by_idx(): can't get object type
    major: Symbol table
    minor: Inappropriate type
  #001: ../../src/H5Gdeprec.c line 1144 in H5G_obj_get_type_by_idx(): can't locate type
    major: Symbol table
    minor: Object not found
  #002: ../../src/H5Gstab.c line 1219 in H5G__stab_get_type_by_idx(): index out of bound
    major: Invalid arguments to routine
    minor: Inappropriate type
*** Checking HDF5 dimension scales.
*** Creating simple dimension scales file...ok.
*** Checking that simple dimscale file can be read...ok.
*** Creating simple dimension scales file with lots of datasets...ok.
*** Creating a file with an unlimited dimension scale...ok.
*** Checking that unlimited dimscale file can be read...Sorry! Unexpected result, ../../h5_test/tst_h_dimscales.c, line: 387
FAIL tst_h_dimscales (exit status: 2)
```
### Steps to reproduce the behavior

See https://copr.fedorainfracloud.org/coprs/orion/hdf5110/build/172669/

### LINKED ISSUE #363: tst_h_dimscales with hdf5-1.10
> Please provide as much of the following information as you can, as applicable to the issue being reported.  Naturally, not all information is relevant to every issue, but the more information we have to start, the better!

### Environment Information
gcc-6.3.0
zlib-1.2.11
hdf5-1.10.0-patch1
I am doing a serial installation (no MPI).

64-bit, Linux, debian Jessie (fully upgraded)
Using autotools (attached config.log)

[config.log.txt](https://github.com/Unidata/netcdf-c/files/776251/config.log.txt)

### Summary of Issue
After having compiled the netcdf-4 C library I issue `make check` and then this test fails:
`FAIL: tst_h_dimscales`
All other tests succeed.

For completeness sake, I have attached the log for the test:

[tst_h_dimscales.log.txt](https://github.com/Unidata/netcdf-c/files/776257/tst_h_dimscales.log.txt)

### Steps to reproduce the behavior
The `config.log` should provide enough details. A "simple" `../configure ...`, `make`, `make check`.

### LINKED ISSUE #505: memory overflow reported in tst_h_dimscales
I built the HEAD with gcc address/memory checking turned on. Turns out there are some memory leaks/overflows.

See #504 for build info.

```
./tst_h_dimscales

*** Checking HDF5 dimension scales.
*** Creating simple dimension scales file...ok.
*** Checking that simple dimscale file can be read...ok.
*** Creating simple dimension scales file with lots of datasets...ok.
*** Creating a file with an unlimited dimension scale...ok.
=================================================================
==22535==ERROR: AddressSanitizer: stack-buffer-overflow on address 0x7ffe37d95610 at pc 0x557a8a8165d8 bp 0x7ffe37d94fa0 sp 0x7ffe37d94f90
WRITE of size 8 at 0x7ffe37d95610 thread T0
    #0 0x557a8a8165d7 in alien_visitor /home/ed/tmp/netcdf-c/h5_test/tst_h_dimscales.c:25
    #1 0x7fbc424f0ed7 in H5DSiterate_scales /home/ed/Downloads/hdf5-1.10.1/hl/src/H5DS.c:1357
    #2 0x557a8a81b104 in main /home/ed/tmp/netcdf-c/h5_test/tst_h_dimscales.c:437
    #3 0x7fbc4195c3f0 in __libc_start_main (/lib/x86_64-linux-gnu/libc.so.6+0x203f0)
    #4 0x557a8a816309 in _start (/home/ed/tmp/netcdf-c/h5_test/tst_h_dimscales+0x2309)

Address 0x7ffe37d95610 is located in stack of thread T0 at offset 32 in frame
    #0 0x557a8a817697 in main /home/ed/tmp/netcdf-c/h5_test/tst_h_dimscales.c:118

  This frame has 12 object(s):
    [32, 36) 'visitor_data' <== Memory access at offset 32 partially overflows this variable
    [96, 104) 'num_obj'
    [160, 168) 'i'
    [224, 232) 'num_obj'
    [288, 296) 'dims'
    [352, 360) 'maxdims'
    [416, 432) 'dims'
    [480, 504) 'dims'
    [544, 568) 'max_dims'
    [608, 4608) 'var1_datasetid'
    [4640, 4896) 'obj_name'
    [4928, 5184) 'label'
HINT: this may be a false positive if your program uses some custom stack unwind mechanism or swapcontext
      (longjmp and C++ exceptions *are* supported)
SUMMARY: AddressSanitizer: stack-buffer-overflow /home/ed/tmp/netcdf-c/h5_test/tst_h_dimscales.c:25 in alien_visitor
Shadow bytes around the buggy address:
  0x100046faaa70: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x100046faaa80: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x100046faaa90: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x100046faaaa0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
  0x100046faaab0: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 f1 f1
=>0x100046faaac0: f1 f1[04]f4 f4 f4 f2 f2 f2 f2 00 f4 f4 f4 f2 f2
  0x100046faaad0: f2 f2 00 f4 f4 f4 f2 f2 f2 f2 00 f4 f4 f4 f2 f2
  0x100046faaae0: f2 f2 00 f4 f4 f4 f2 f2 f2 f2 00 f4 f4 f4 f2 f2
  0x100046faaaf0: f2 f2 00 00 f4 f4 f2 f2 f2 f2 00 00 00 f4 f2 f2
  0x100046faab00: f2 f2 00 00 00 f4 f2 f2 f2 f2 00 00 00 00 00 00
  0x100046faab10: 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00 00
Shadow byte legend (one shadow byte represents 8 application bytes):
  Addressable:           00
  Partially addressable: 01 02 03 04 05 06 07 
  Heap left redzone:       fa
  Heap right redzone:      fb
  Freed heap region:       fd
  Stack left redzone:      f1
  Stack mid redzone:       f2
  Stack right redzone:     f3
  Stack partial redzone:   f4
  Stack after return:      f5
  Stack use after scope:   f8
  Global redzone:          f9
  Global init order:       f6
  Poisoned by user:        f7
  Container overflow:      fc
  Array cookie:            ac
  Intra object redzone:    bb
  ASan internal:           fe
  Left alloca redzone:     ca
  Right alloca redzone:    cb
==22535==ABORTING> 
```

### PR DESCRIPTION
**Pull request by edhartnett**
Fixes #505.
Fixes #363.
Fixes #244.

Simple fix - passed parameter was size int, but was being cast to int64. As a result, there was a memory overflow which may blow away the loop counter variable.

Hilarious confusion then ensued.

### CODE REVIEW

File: h5_test/tst_h_dimscales.c [hunk 1]
```diff
@@ -379,7 +379,7 @@ main()
       if ((grpid = H5Gopen(fileid, GRP_NAME)) < 0) ERR;
 
       /* Loop through datasets to find variables. */
-      if (H5Gget_num_objs(grpid, &num_obj) < 0) ERR;
+      if (H5Gget_num_objs(grpid, (hsize_t *)&num_obj) < 0) ERR;
```
Comment (edhartnett) [inline h5_test/tst_h_dimscales.c]: Made this change to eliminate warning. H5Gget_num_objs does not expect a volatile.

**Comment by WardF** [comment]
Super duper, and obviates a branch Iw as working on locally (regarding the volatile).  Can you review and accept the CLA?  You should only need to do that once. It's a Unidata thing and applies across all the repositories under `Unidata/`, it isn't specific to netCDF (or you, hah!).  Thanks!

**Comment by edhartnett** [comment]
I have signed the CLA.

**Comment by WardF** [comment]
Weird, I'm seeing a strange error message for this PR over at Travis:

* https://travis-ci.org/Unidata/netcdf-c/requests

I've just reached out to travis support and will see what they have to say as far as this goes.

**Comment by WardF** [comment]
@edhartnett I got a response from the travis CI folks:

> Thanks for reaching out and I'm sorry for the troubles.
> 
> Looking at the contributor's account (edhartnett), it seems that they have never logged into > Travis CI and this is why we are currently blocking their requests. Sorry for the inconvenience.
>
> To confirm, do you think you could have them try to log into Travis CI at least once at https://travis-ci.org/ and push a new commit?
>
> Please let us know the result and we would be happy to have another look if it's still not working.
>
> Thanks in advance and have a great day!

It looks like the checks passed so maybe you already did this, but I thought I'd pass it along anyways.ooo

**Comment by edhartnett** [comment]
I have logged into travis - let's see if that eliminates the problem next time.

**Pull request merged**